sql/inspect: add checkpointing tracking to INSPECT

spilchen opened the pull request:
Previously, INSPECT jobs did not support checkpointing, meaning job restarts would reprocess all data from the beginning. This change adds span based checkpointing to allow jobs to resume from their last completed spans.

Progress is tracked using jobfrontier to store completed spans. On job restart, the resumer loads completed spans and filters them from the work to be done.

Closes #148297

Release note: None

Epic: CRDB-30356

rafiss reviewed (changes requested):
looks good! just had small comments

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154276#-:-OaM5wIN-y7m2NsSQq_M:b-31cul3)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/inspect/inspect_job.go` line 71 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154276#-OaL7Re6B74DoE4lvpVw:-OaL7Re76cHlqNJSYhCN:b-cjq8hp) ([raw file](https://github.com/cockroachdb/cockroach/blob/606563e3a09d130c32e0ab0466add90a4a6ecda7/pkg/sql/inspect/inspect_job.go#L71)):*
> ```Go
> 	// If all spans are completed, job is done
> 	if len(filteredSpans) == 0 {
> 		log.Dev.Infof(ctx, "All spans already completed, INSPECT job finished")
> ```

nit: use lower case for log messages 
___
*[`pkg/sql/inspect/inspect_job.go` line 238 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154276#-OaLmqJD3FAfaDYKClFX:-OaLmqJD3FAfaDYKClFY:b-31orhk) ([raw file](https://github.com/cockroachdb/cockroach/blob/606563e3a09d130c32e0ab0466add90a4a6ecda7/pkg/sql/inspect/inspect_job.go#L238)):*
> ```Go
> 
> // setupProgressTrackingAndFilter initializes progress tracking and returns
> // it, along with a cleanup function and the completed spans to filter.
> ```

i didn't follow "and the completed spans to filter." isn't this function already doing the filtering?
___
*[`pkg/sql/inspect/inspect_job.go` line 253 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154276#-OaLmgBr6A4pxEC0G9Wc:-OaLmgBsCs6Cx_N3YErp:b195v9s) ([raw file](https://github.com/cockroachdb/cockroach/blob/606563e3a09d130c32e0ab0466add90a4a6ecda7/pkg/sql/inspect/inspect_job.go#L253)):*
> ```Go
> 		return nil, nil, nil, err
> 	}
> 	filteredSpans := c.filterCompletedSpans(pkSpans, completedSpans)
> ```

nit: perhaps naming it `remainingSpans` would be more clear than `filteredSpans`
___
*[`pkg/sql/inspect/progress.go` line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154276#-OaM3Hm1CGp4pM6mipgV:-OaM3Hm1CGp4pM6mipgW:b-obht53) ([raw file](https://github.com/cockroachdb/cockroach/blob/606563e3a09d130c32e0ab0466add90a4a6ecda7/pkg/sql/inspect/progress.go#L27)):*
> ```Go
> 	"github.com/cockroachdb/errors"
> 	pbtypes "github.com/gogo/protobuf/types"
> 	"golang.org/x/sync/errgroup"
> ```

instead of `errgroup`, I recommend using `ctxgroup` (from pkg/util/ctxgroup). The docs there explain why -- it makes the context handling much more clear.
___
*[`pkg/sql/inspect/progress.go` line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154276#-OaM4ps-1NprtGrZ5s8i:-OaM4ps-1NprtGrZ5s8j:b533dnr) ([raw file](https://github.com/cockroachdb/cockroach/blob/606563e3a09d130c32e0ab0466add90a4a6ecda7/pkg/sql/inspect/progress.go#L85)):*
> ```Go
> 		}
> 
> 		// Extract just the spans (we don't care about timestamps for INSPECT)
> ```

super nit: this is not a blocking comment, but I've noticed in this PR and some of the prior ones that some comments are missing punctuation at the end. no need to go back and change all of them, but just wanted to call it out since sometimes the punctuation actually does improve clarity.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/inspect: add checkpointing to INSPECT jobs

Previously, INSPECT jobs did not support checkpointing, meaning job restarts
would reprocess all data from the beginning. This change adds span based
checkpointing to allow jobs to resume from their last completed spans.

Progress is tracked using jobfrontier to store completed spans. On job
restart, the resumer loads completed spans and filters them from the work to
be done.

Closes #148297

Release note: None

Epic: CRDB-30356

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154276)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/inspect/inspect_job.go` line 238 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154276#-OaLmqJD3FAfaDYKClFX:-OaRx6S85dlV9OkHEry8:bcrp4sk) ([raw file](https://github.com/cockroachdb/cockroach/blob/606563e3a09d130c32e0ab0466add90a4a6ecda7/pkg/sql/inspect/inspect_job.go#L238)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i didn't follow "and the completed spans to filter." isn't this function already doing the filtering?
</blockquote></details>

You are correct. It does the filtering. I'll update the comment.
___
*[`pkg/sql/inspect/inspect_job.go` line 253 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154276#-OaLmgBr6A4pxEC0G9Wc:-OaRxAAI6-BNB_NapdZJ:b-h76dzt) ([raw file](https://github.com/cockroachdb/cockroach/blob/606563e3a09d130c32e0ab0466add90a4a6ecda7/pkg/sql/inspect/inspect_job.go#L253)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: perhaps naming it `remainingSpans` would be more clear than `filteredSpans`
</blockquote></details>

Yes, that's a clearer name.
___
*[`pkg/sql/inspect/progress.go` line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154276#-OaM3Hm1CGp4pM6mipgV:-OaRyFDI9EO0sSOLHSrv:bqihut1) ([raw file](https://github.com/cockroachdb/cockroach/blob/606563e3a09d130c32e0ab0466add90a4a6ecda7/pkg/sql/inspect/progress.go#L27)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

instead of `errgroup`, I recommend using `ctxgroup` (from pkg/util/ctxgroup). The docs there explain why -- it makes the context handling much more clear.
</blockquote></details>

Done. Thanks for the tip. I wonder if a lint rule would have caught this earlier.
___
*[`pkg/sql/inspect/progress.go` line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154276#-OaM4ps-1NprtGrZ5s8i:-OaRyUXaEEoIjGCr083l:b-nky66d) ([raw file](https://github.com/cockroachdb/cockroach/blob/606563e3a09d130c32e0ab0466add90a4a6ecda7/pkg/sql/inspect/progress.go#L85)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: this is not a blocking comment, but I've noticed in this PR and some of the prior ones that some comments are missing punctuation at the end. no need to go back and change all of them, but just wanted to call it out since sometimes the punctuation actually does improve clarity.
</blockquote></details>

Okay, thanks. I'll try to keep this in mind. I've updated the new comments I have added in this PR.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154276#-:-OaaHVi57F34e3IwQalz:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen)



<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+
